go-client/client/client.go

```diff
@@ -0,0 +1,262 @@
+func NewClient(ctx context.Context, host string, port string, scriptLanguage string) (*Client, error) {
+	grpcChannel, err := grpc.Dial(host+":"+port, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, err
+	}
+
+	client := &Client{grpcChannel: grpcChannel, tables: make(map[fieldId]*TableHandle)}
+
+	client.sessionStub, err = newSessionStub(ctx, client)
+	if err != nil {
+		client.Close()
+		return nil, err
+	}
+
+	client.tableStub, err = newTableStub(client)
+	if err != nil {
+		client.Close()
+		return nil, err
+	}
+
+	client.consoleStub, err = newConsoleStub(ctx, client, scriptLanguage)
+	if err != nil {
+		client.Close()
+		return nil, err
+	}
+
+	client.flightStub, err = newFlightStub(client, host, port)
+	if err != nil {
+		client.Close()
+		return nil, err
+	}
+
+	client.inputTableStub = newInputTableStub(client)
+
+	client.appStub = newAppStub(client)
+
+	return client, nil
+}
+
+// Checks if the client is closed, i.e. it can no longer perform operations on the server.
+func (client *Client) Closed() bool {
+	return client.grpcChannel == nil
+}
+
+type FetchOption int
+
+const (
+	FetchOnce      FetchOption = iota // Fetches the list of tables once and then returns.
+	FetchRepeating FetchOption = iota // Starts up a background thread to continually update the list of tables as changes occur.
+)
+
+// Fetches the list of tables from the server.
+// This allows the client to see the list of named global tables on the server,
+// and thus allows it to open them using OpenTables.
+// Tables created in scripts run by the current client are immediately visible and do not require a FetchTables call.
+func (client *Client) FetchTables(ctx context.Context, opt FetchOption) error {
+	if client.Closed() {
+		return ErrClosedClient
+	}
+
+	err := client.listFields(ctx, func(update *applicationpb2.FieldsChangeUpdate) {
+		client.handleFieldChanges(update)
+	})
+	if err != nil {
+		return err
+	}
+
+	if opt == FetchOnce {
+		client.appStub.cancelFunc()
+	}
+
+	return nil
+}
+
+// Returns a list of the (global) tables that can be opened with OpenTable.
+// This can be updated using FetchTables.
+func (client *Client) ListOpenableTables() []string {
+	client.tablesLock.Lock()
+	defer client.tablesLock.Unlock()
+
+	var result []string
+	for id := range client.tables {
+		if id.appId == "scope" {
+			result = append(result, id.fieldName)
+		}
+	}
+	return result
+}
+
+func (client *Client) newTicketNum() int32 {
+	client.nextTicket += 1
+	if client.nextTicket <= 0 {
+		// TODO:
+		panic("out of tickets")
+	}
+
+	return client.nextTicket
+}
+
+func (client *Client) newTicket() ticketpb2.Ticket {
+	id := client.newTicketNum()
+
+	return client.makeTicket(id)
+}
+
+func (client *Client) makeTicket(id int32) ticketpb2.Ticket {
+	bytes := []byte{'e', byte(id), byte(id >> 8), byte(id >> 16), byte(id >> 24)}
+
+	return ticketpb2.Ticket{Ticket: bytes}
+}
```

Comment:
If this is only called with `id := client.newTicketNum()`, consider condensing this into `newTicket`

Reply:
It is used in ImportTable because it needs the ticket number and the ticket itself separately. I could either keep `makeTicket`, or make a new method that returns both the ticket and the number used to create it. Both approaches feel more or less equivalent.

Reply:
this is fine, if it is used elsewhere.